### PR TITLE
Remove default_args pattern + added get_current_context() use for Core Airflow example DAGs

### PR DESCRIPTION
**Pull request by josh-fell**
Related to #10285

- Removed or refactored the `default_args` pattern where necessary as requested by @ashb (i.e. removed a separated `default_args` declaration for deference for declaration as part of the `DAG` object)
- Updated example DAG files such that `xcom_pull()` calls use an operator's `.output` property as well access of `TaskInstance` objects from context to use `get_current_context()` function
- Added comments to which task dependencies, if any, are handled and/or created via `XComArgs` for transparency

An detailed summary of all changes made as part of this PR can be found below:
| DAG File | Converted `xcom_pull()`? | Other Updates? | Comments |
| ---------| ------------------------ | -------------- | -------- |
| airflow/example_dags/example_bash_operator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_branch_datetime_operator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_branch_day_of_week_operator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_branch_operator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_branch_python_dop_operator_3.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_dag_decorator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_kubernetes_executor_config.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_kubernetes_executor.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_passing_params_via_test_command.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_python_operator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_short_circuit_operator.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_skip_dag.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_trigger_controller_dag.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_trigger_target_dag.py | No | Yes | Removed unneeded `default_args` pattern. </br></br>Updated to use `get_current_context()`.|
| airflow/example_dags/example_xcom.py | Yes | Yes | Included examples for context retrieval using `get_current_context()` based on previous PR comments to show new and old means of `XCom` access.<br/><br/>Removed explicit task dependencies that are created via `XComArgs`. <br/><br/>Removed unneeded `default_args` pattern. |
| airflow/example_dags/example_xcomargs.py | No | Yes | Removed unneeded `default_args` pattern. |
| airflow/example_dags/tutorial_etl_dag.py | No | No | Not updated based on earlier comment to keep this file unchanged during initial PR. |
| airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py | No | Yes | Removed unneeded `default_args` pattern. </br></br> Moved `import json` statement to inside `extract()` function to prevent failures.  |
| airflow/example_dags/tutorial_taskflow_api_etl.py | No | Yes | Removed unneeded `default_args` pattern. |

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/example_dags/example_trigger_target_dag.py
Comment (ashb) [inline airflow/example_dags/example_trigger_target_dag.py]:
```suggestion
def run_this_func(dag_run):
    """
    Print the payload "message" passed to the DagRun conf attribute.

    """
    print(f"Remotely received value of {dag_run.conf['message']} for key=message")
```

**Review by ashb — CHANGES_REQUESTED** [review]
See https://github.com/apache/airflow/pull/16872#pullrequestreview-701813622

File: airflow/example_dags/example_trigger_target_dag.py
Comment (josh-fell) [reply airflow/example_dags/example_trigger_target_dag.py]: 👍

**Review by potiuk — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
@ashb - waiting for your approval/merge

File: airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py
Comment (uranusjr) [inline airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py:22]: Why move this?

File: airflow/example_dags/example_trigger_target_dag.py
Comment (ashb) [inline airflow/example_dags/example_trigger_target_dag.py]:
```suggestion
    :type dag_run: DagRun
```

File: airflow/example_dags/example_xcom.py
Comment (ashb) [inline airflow/example_dags/example_xcom.py]: I'm not really sure we need to show `get_current_context()` in example DAGs at all -- this feels like an anti-pattern here or an advanced use thing only, and people copy examples more than we might like.

File: airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py
Comment (josh-fell) [reply airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py:22]: Unfortunately the "extract" task fails without it.
```[2021-08-06, 13:26:50 UTC] {process_utils.py:135} INFO - Executing cmd: /tmp/venvys6_yc45/bin/python /tmp/venvys6_yc45/script.py /tmp/venvys6_yc45/script.in /tmp/venvys6_yc45/script.out /tmp/venvys6_yc45/string_args.txt
[2021-08-06, 13:26:50 UTC] {process_utils.py:139} INFO - Output:
[2021-08-06, 13:26:50 UTC] {process_utils.py:143} INFO - Traceback (most recent call last):
[2021-08-06, 13:26:50 UTC] {process_utils.py:143} INFO -   File "/tmp/venvys6_yc45/script.py", line 40, in <module>
[2021-08-06, 13:26:50 UTC] {process_utils.py:143} INFO -     res = extract(*arg_dict["args"], **arg_dict["kwargs"])
[2021-08-06, 13:26:50 UTC] {process_utils.py:143} INFO -   File "/tmp/venvys6_yc45/script.py", line 37, in extract
[2021-08-06, 13:26:50 UTC] {process_utils.py:143} INFO -     order_data_dict = json.loads(data_string)
[2021-08-06, 13:26:50 UTC] {process_utils.py:143} INFO - NameError: name 'json' is not defined
```

File: airflow/example_dags/example_xcom.py
Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: I think having an another example using the `get_current_context()` function outside of the [Concepts documentation](https://airflow.apache.org/docs/apache-airflow/2.0.0/concepts.html#accessing-current-context) could be beneficial for those who really like using task-decorated functions. Maybe rewriting the example that more closely aligns with the concept docs plus some detailed comments could be acceptable. WDYT?

I do also agree that if there is a doubt about displaying an example we shouldn't even bother given all the blind copy/paste happening. Certainly some other updates could be made to examples which border on anti-patterns (e.g. `days_ago()` in `start_date`). I feel the rewrite would alleviate doubt here though.

Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: @ashb I updated the example when using `get_current_context()`. If you still have some concerns, happy to nuke showing this method from the example DAG so we can push these updates forward.  Thanks for your feedback as always!

**Review by uranusjr — APPROVED** [review]

File: airflow/example_dags/example_xcom.py
Comment (ephraimbuddy) [inline airflow/example_dags/example_xcom.py]: I think that we shouldn't use PythonOperator at all for this example if we want to use the TaskFlow API because TaskFlowAPI is an abstraction of the PythonOperator. Using both the TaskFlow API and PythonOperator in the same example doesn't look great to me.

Comment (ephraimbuddy) [reply airflow/example_dags/example_xcom.py]:
```python

from airflow import DAG
from airflow.utils.dates import days_ago

value_1 = [1, 2, 3]
value_2 = {'a': 'b'}

dag = DAG(
    'example_xcom',
    schedule_interval="@once",
    start_date=days_ago(2),
    tags=['example'],
)

@dag.task()
def push():
    """Pushes an XCom without a specific target"""
    return value_1


@dag.task()
def push_by_returning():
    """Pushes an XCom without a specific target, just by returning it"""
    return value_2


@dag.task()
def puller(data1, data2):
    """Pull all previously pushed XComs and check if the pushed values match the pulled values."""

    # get value_1
    pulled_value_1 = data1
    if pulled_value_1 != value_1:
        raise ValueError(f'The two values differ {pulled_value_1} and {value_1}')

    # get value_2
    pulled_value_2 = data2
    if pulled_value_2 != value_2:
        raise ValueError(f'The two values differ {pulled_value_2} and {value_2}')

    # get both value_1 and value_2
    pulled_value_1, pulled_value_2 = data1, data2
    if pulled_value_1 != value_1:
        raise ValueError(f'The two values differ {pulled_value_1} and {value_1}')
    if pulled_value_2 != value_2:
        raise ValueError(f'The two values differ {pulled_value_2} and {value_2}')


puller(push(), push_by_returning())

```

Comment (ephraimbuddy) [reply airflow/example_dags/example_xcom.py]: If we want to push with a specific ID:

```python
from airflow import DAG
from airflow.utils.dates import days_ago

value_1 = [1, 2, 3]
value_2 = {'a': 'b'}

dag = DAG(
    'example_xcom',
    schedule_interval="@once",
    start_date=days_ago(2),
    tags=['example'],
)

@dag.task()
def push(**kwargs):
    """Pushes an XCom without a specific target"""
    kwargs['ti'].xcom_push(key='value from pusher 1', value=value_1)


@dag.task()
def push_by_returning():
    """Pushes an XCom without a specific target, just by returning it"""
    return value_2


@dag.task()
def puller(data1, data2, **kwargs):
    """Pull all previously pushed XComs and check if the pushed values match the pulled values."""
    ti = kwargs['ti']
    # get value_1
    pulled_value_1 = ti.xcom_pull(key=None, task_ids='push')
    if pulled_value_1 != value_1:
        raise ValueError(f'The two values differ {pulled_value_1} and {value_1}')

    # get value_2
    pulled_value_2 = data2
    if pulled_value_2 != value_2:
        raise ValueError(f'The two values differ {pulled_value_2} and {value_2}')

    # get both value_1 and value_2
    pulled_value_1, pulled_value_2 = ti.xcom_pull(key=None, task_ids=['push', 'push_by_returning'])
    if pulled_value_1 != value_1:
        raise ValueError(f'The two values differ {pulled_value_1} and {value_1}')
    if pulled_value_2 != value_2:
        raise ValueError(f'The two values differ {pulled_value_2} and {value_2}')


puller(push(), push_by_returning())
```

Comment (ephraimbuddy) [reply airflow/example_dags/example_xcom.py]: I think this is simple, easy to understand, and demonstrated the power of Taskflow API

Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: @ephraimbuddy @ashb @uranusjr What do you think about creating a new example DAG that strictly shows `XCom` push/pull with the Taskflow API? I agree with Ephraim and feel we should have an example for Taskflow API here -- albeit the original thought was to demonstrate using the `get_current_context()` method -- especially if see (or want) the community to adopt this style. (@ephraimbuddy @uranusjr love to get your thoughts on showcasing that method as well especially with some concerns on blind copy/paste that I agree with @ashb).

Comment (uranusjr) [reply airflow/example_dags/example_xcom.py]: What would the existing examples look like if we add the new example? I think it’d be a good idea to have separate examples demostrating the Taskflow API and the “traditional” context manager or `Operator(dag=dag)` approach, since they do have slightly different best practices.

Comment (ephraimbuddy) [reply airflow/example_dags/example_xcom.py]: For me, I think we shouldn't show any example with PythonOperator. We should instead use TaskFlow API. However, whenever we have BashOperator and the likes, we should use context manager or `Operator(dag=dag)`. TaskFlow can be used inside context manager, for cases where we have a combination of PythonOperator and Other Operators:

```python
with dag:
    task1 = BashOperator(task_id='myid', bash_command='sleep 1')
    
    @task
    def mypythontask(**kwargs):
        return 'xcom'

task1 >> mypythontask()
```
I feel we should stop using PythonOperator in examples.

Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: I do have #9415 assigned to me as well to update all of the example DAGs to use the Taskflow API so eventually we'll be getting to a point where `PythonOperator` and `PythonVirtualEnvOperator` (and perhaps `TaskGroup` context manager if it makes sense in the example) are not used in the examples or documentation.  

Initially the thought was to tackle #10285 first (this PR and others to use the `.output` property and refactoring the `default_args` pattern mainly) then go back to tackle #9415 since I was already deep into updating 86 files encompassing #10285.  

Seems like it makes sense to remove the Taskflow API and `get_current_context()` examples in `example_xcom.py` and then continue/revisit this discussion in a later PR.  WDYT?

Comment (ephraimbuddy) [reply airflow/example_dags/example_xcom.py]: Sounds good...

File: airflow/example_dags/example_xcom.py
Comment (ashb) [inline airflow/example_dags/example_xcom.py]: No, this is wrong, and I think this much simpler way will work:

```suggestion
@task
def push_with_taskflow(ti, val=value_2):
    """
    Pushes an XCom from within a task-decorated function aka using the Taskflow API.
    """
    ti.xcom_push(key='other value from pusher 1', value=val)
```

(But could you test it please)

Comment (ashb) [reply airflow/example_dags/example_xcom.py]: i.e. I'm not sure _when_ `get_current_context()` is ever actually needed (at least not within an operator or a task function)

Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: > No, this is wrong, and I think this much simpler way will work:
> 
> (But could you test it please)

Tested this out. Just needed to make `ti` an optional arg -- otherwise DAG parsing would throw an exception -- and it worked just fine.  Seems to make the use of `get_current_context()` more nebulous.

File: airflow/example_dags/example_xcom.py
Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: Reverted Taskflow API and `get_current_context()` changes in the "example_xcom" DAG.

File: airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py
Comment (ephraimbuddy) [inline airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py:31]: Do we have a place we explained about default_args being passed to all tasks? Probably we should have one example

**Review by ephraimbuddy — APPROVED** [review]
LGTM

File: airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py
Comment (josh-fell) [reply airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py:31]: @ephraimbuddy We still have an explanation in the `tutorial.py` example DAG.

```python
# [START default_args]
# These args will get passed on to each operator
# You can override them on a per-task basis during operator initialization
default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email': ['airflow@example.com'],
    'email_on_failure': False,
    'email_on_retry': False,
    'retries': 1,
    'retry_delay': timedelta(minutes=5),
    # 'queue': 'bash_queue',
    # 'pool': 'backfill',
    # 'priority_weight': 10,
    # 'end_date': datetime(2016, 1, 1),
    # 'wait_for_downstream': False,
    # 'dag': dag,
    # 'sla': timedelta(hours=2),
    # 'execution_timeout': timedelta(seconds=300),
    # 'on_failure_callback': some_function,
    # 'on_success_callback': some_other_function,
    # 'on_retry_callback': another_function,
    # 'sla_miss_callback': yet_another_function,
    # 'trigger_rule': 'all_success'
}
# [END default_args]
```

When I go through the example DAGs again for Taskflow API updates, I'll see what applicable `default_args` use I can sprinkle in too.

Comment (josh-fell) [reply airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py:31]: Whether or not that is a _good_ example is another question though. There is a lot of blind copy/paste by users in DAG files and maybe it makes sense to update the documentation to list these out in an explicit table rather than hiding in DAG code.

We've been trying to improve the `default_args` use (e.g. not specifying default `default_args` and/or moving useful/applicable `default_args` assignment directly to the `DAG()` instantiation). I could definitely add this table as part of the upcoming PR for example DAGs if you agree on this approach.

Comment (ephraimbuddy) [reply airflow/example_dags/tutorial_taskflow_api_etl_virtualenv.py:31]: Makes sense @josh-fell. Thanks for that.

File: airflow/example_dags/example_xcom.py
Comment (josh-fell) [reply airflow/example_dags/example_xcom.py]: To close the gap on this comment, I've gone ahead and removed any TaskFlow API-related changes (reference [this comment](https://github.com/apache/airflow/pull/16866#discussion_r689944981)).  We can discuss later when I tackle another example DAG issue but passing in the `ti` object as optional works just fine and I'll incorporate this change then.

**Review by ashb — APPROVED** [review]

**Pull request merged**
